onnxruntime/python/tools/qnn/add_trans_cast.py

```diff
@@ -0,0 +1,300 @@
+        for output in sorted_nodes[start].output:
+            if output in deps_to_nodes:
+                for node_idx in deps_to_nodes[output]:
+                    deps_count[node_idx] = deps_count[node_idx] - 1
+                    if deps_count[node_idx] == 0:
+                        sorted_nodes.append(graph.node[node_idx])
+                        end = end + 1
+        start = start + 1
+
+    assert (end == len(graph.node)), "Graph is not a DAG"
+    graph.ClearField('node')
+    graph.node.extend(sorted_nodes)
+
+def get_value_from_qnn_code_line(line_content):
+    value = line_content.split("=")
+    assert (len(value) == 2), "Error: Wrong format encountered."
+    value = value[1].split(",")
+    assert (len(value) == 2), "Error: Wrong format encountered."
+    return value[0].replace(" ", "").replace("\"", "")
+
+class qnn_tensor_struct:
+    def __init__(self):
+        self.name = ""
+        self.qnn_data_type = ""
+        self.onnx_data_type = TensorProto.FLOAT
+        self.dim_alias = ""
+        self.dim = []
+
+def qnn_data_type_str_to_onnx_data_type(qnn_data_type_str):
+    if qnn_data_type_str == "QNN_DATATYPE_UFIXED_POINT_8" or qnn_data_type_str == "QNN_DATATYPE_UINT_8":
+        return TensorProto.UINT8
+    elif qnn_data_type_str == "QNN_DATATYPE_UFIXED_POINT_16" or qnn_data_type_str == "QNN_DATATYPE_UINT_16":
+        return TensorProto.UINT16
+    elif qnn_data_type_str == "QNN_DATATYPE_UFIXED_POINT_32" or qnn_data_type_str == "QNN_DATATYPE_UINT_32":
+        return TensorProto.UINT32
+    elif qnn_data_type_str == "QNN_DATATYPE_UINT_64":
+        return TensorProto.UINT64
+    elif qnn_data_type_str == "QNN_DATATYPE_UFIXED_POINT_8":
+        return TensorProto.UINT8
+    elif qnn_data_type_str == "QNN_DATATYPE_FIXED_POINT_8" or qnn_data_type_str == "QNN_DATATYPE_INT_8":
+        return TensorProto.INT8
+    elif qnn_data_type_str == "QNN_DATATYPE_FIXED_POINT_16" or qnn_data_type_str == "QNN_DATATYPE_INT_16":
+        return TensorProto.INT16
+    elif qnn_data_type_str == "QNN_DATATYPE_FIXED_POINT_32" or qnn_data_type_str == "QNN_DATATYPE_INT_32":
+        return TensorProto.INT32
+    elif qnn_data_type_str == "QNN_DATATYPE_INT_64":
+        return TensorProto.INT64
+    elif qnn_data_type_str == "QNN_DATATYPE_FLOAT_16":
+        return TensorProto.FLOAT16
+    elif qnn_data_type_str == "QNN_DATATYPE_FLOAT_32":
+        return TensorProto.FLOAT
+    elif qnn_data_type_str == "QNN_DATATYPE_BOOL_8 ":
+        return TensorProto.BOOL
+    else:
+        return TensorProto.UNDEFINED
+
+def parse_dims(line_content):
+    value = line_content.split("=")
+    assert (len(value) == 2), "Error: Wrong format encountered."
+    value = value[1].replace(" ", "").split(";")
+    assert (len(value) == 2), "Error: Wrong format encountered."
+    shape = value[0].replace("{", "").replace("}", "").split(",")
+    return [eval(item) for item in shape]
+
+
+def parse_qnn_cpp_file(qnn_cpp_file, qnn_input_output_tensor_dic):
```

Comment:
Can we only parse the model.cpp file? The qnn-onnx-converter also outputs a model_net.json file. Getting this information from the generated model_net.json file would probably be simpler and more reliable.

`model_json["graph"]["tensors"]` is a dictionary of tensor information that includes type, data_type, dims, etc.

Reply:
You are right. I didn't notice they output a json file now. That will make it a lot easier and reliable.

---
In reply to: [1235594519](https://github.com/microsoft/onnxruntime/pull/16423#discussion_r1235594519) [](ancestors = 1235594519)